File: R/euler_31.R
# I am sure there is some clever way to make this functional,
# but I'm not enough of an R wizard to do that without a lot of
# reading. So this is a straight translation of the Julia version
count.splits <- function(amount, coins){
	scoins <- sort(coins)

	largest <- tail(scoins, 1L)

	rest <- amount
	count <- 0L

	while (rest > 0) {
		if (length(scoins) > 1L) {
			count <- count + count.splits(rest, head(scoins, -1L))
		}

		rest <- rest - largest
	}

	if (rest == 0) {
		count <- count + 1
	}

	count
}


print("# of ways to pay £2:")
print(count.splits(200, c(200, 100, 50, 20, 10, 5, 2, 1)))

File: julia/euler_31.jl



function count_splits(amount, coins)
	# we want to check them in order
	s_coins = sort(coins)

	# current coin
	largest = s_coins[end]

	rest = amount
	count = 0

	# * this could be done more efficiently by checking
	#   length==0 before the loop, but I think this version 
	#   is prettier
	# * we could improve this even further by using memoization, but
	#   that's overkill in this context
	while rest > 0.0
		# more coins to check
		if length(coins) > 1
			count += count_splits(rest, s_coins[1:(end-1)])
		end
		rest -= largest
	end

	# our current coin actually fits as well
	if rest == 0.0
		count += 1
	end

	count
end
